### PR TITLE
Autoformat Java examples

### PR DESCRIPTION
**Pull request by myyk**
* [x] Need to merge #3903 first because it's built directly on top of it to reduce conflicts.

Change 2/3 for #3829

* I updated the `PalantirFormatModule.scala` so that it didn't create a ton of noise when there's actually no issues such as no java code changing. I think this is a good thing to do, but I'm not 100% since some users might feel this change in some way.
* Bugfix: The Github Action now fails on any error instead of only `./mill -i __.fix --check` (was using `;` instead of `&&`)
* Added Palantir formatting into the Github Action
* Removed a lot of cruft so that the example java modules could use the `PalantirFormatModule`.

### CODE REVIEW

File: example/javalib/linting/2-checkstyle/src/InputWhitespaceCharacters.java [hunk 1]
```diff
@@ -12,8 +12,10 @@ final class InputWhitespaceCharacters {
    * @return hack
    * @throws java.lang.Exception abc
    **/
-  int test1(int badFormat1, int badFormat2,
-  		final int badFormat3) // violation 'Line contains a tab character.'
+  int test1(
+      int badFormat1,
+      int badFormat2,
+      final int badFormat3) // violation 'Line contains a tab character.'
```
Comment (myyk) [inline example/javalib/linting/2-checkstyle/src/InputWhitespaceCharacters.java:18]: TODO: Need to restore this lint error and find a way to allow it as an exception. (Can't finish today.)

File: example/javalib/linting/3-palantirformat/src/A.java [hunk 2]
```diff
@@ -1,6 +1,6 @@
 public class A {
 
-    public static void main(String[] args) {
+  public static void main(String[] args) {
     System.out.println("hello"); // indentation should be fixed
```
Comment (myyk) [inline example/javalib/linting/3-palantirformat/src/A.java:4]: TODO: Need to restore this lint error and find a way to allow it as an exception. (Can't finish today.)

File: example/javalib/linting/3-palantirformat/src/A.java [hunk 3]
```diff
@@ -1,6 +1,6 @@
...
-    }
+  }
 }
```
Comment (myyk) [inline example/javalib/linting/3-palantirformat/src/A.java:6]: TODO: Need to restore this lint error and find a way to allow it as an exception. (Can't finish today.)

**Comment by myyk** [comment]
@lihaoyi I'm missing some fixes on keeping the linting examples there. Otherwise I think this one is about ready.

File: example/javalib/linting/3-palantirformat/src/A.java (hunk 3 above)
Comment (myyk) [reply example/javalib/linting/3-palantirformat/src/A.java]: Done

File: example/javalib/linting/3-palantirformat/src/A.java (hunk 2 above)
Comment (myyk) [reply example/javalib/linting/3-palantirformat/src/A.java]: Done

File: example/javalib/linting/2-checkstyle/src/InputWhitespaceCharacters.java (hunk 1 above)
Comment (myyk) [reply example/javalib/linting/2-checkstyle/src/InputWhitespaceCharacters.java]: Done

**Comment by myyk** [comment]
@lihaoyi I think this one is ready for a review when you get a chance.

**Review by lihaoyi — APPROVED** [review]
Looks great, I think just need to update the test cases and we're good to go

**Comment by lihaoyi** [comment]
Looks like a bunch of `PalantirFormatModuleTest` tests are failing. I'm guessing those expect their test data to be mis-formatted, so we might need to figure out a way to skip those

**Comment by lihaoyi** [comment]
So PalantirJavaFormat does not provide any builtin flags to ignore files or folders, instead Mill manages the listing of files to pass to the tool.

https://github.com/com-lihaoyi/mill/blob/4c05156ccd8eca733411139b40886af7fcc485dc/scalalib/src/mill/javalib/palantirformat/PalantirFormatModule.scala#L173-L181

So we likely need to add some logic in Mill to allow `PalantirJavaFormatModule` to ignore particular globs, and then use that in Mill's own build (after rebootstrapping) to ignore the files and roll out enforcement internally.

To do this, I suggest we overload `PalantirFormatModule.formatAll` to take an additional optional parameter `excludedGlobs: Seq[String]`. We can then plumb those through to the java-file-listing above and use it to ignore files matching a glob (may need to find a glob-matching implementation online). We'd need to preserve the current `formatAll` signature for binary compatibility, but it can be a dumb forwarder to the new implementation

**Comment by myyk** [comment]
I had actually already implemented skipping some files from the .mill side
but I guess I missed something.

**Comment by myyk** [comment]
The glob solution seems like a good way to go but I'm not sure when I can
do it. Kotlin needs a similar mechanism and I think it would be good to
have all the linter modules have roughly the same interface (where
possible) which they don't currently.

**Comment by lihaoyi** [comment]
Ah I just noticed your skipping code in Mill's own build. Not sure why it's not working

**Comment by lihaoyi** [comment]
Pushed some fixes, let's see if this works

**Comment by myyk** [comment]
Oh interesting. I see you changed the `sources` in `example/build.mill`, is that what added a lot more linting to this? That found a lot of files I missed.

**Comment by lihaoyi** [comment]
@myyk yeah, previously we only used `lintableJavaSources` in `PalantirFormatModule/`, and didnt use it in `ScalafmtModule/` so the Scala `.mill` files were not picked up. Furthermore, `sources` was pointed at `testRepoRoot`, which for the `javalib/` examples contains a copy of the files and not the original ones so formatting them doesn't do anything.

By re-using the name `sources`, both `PalantirFormatModule/` and `ScalafmtModule/` pick up all of them, which is why now the `.mill` files in the Java folders are being picked up. And I re-targeted `sources` directly at `millSourcePath`, so it doesn't get affected by the override of `testRepoRoot` in `ExampleCrossModuleJava`

**Comment by myyk** [comment]
That all makes sense. Oh, I think I fixed some of that in the next change #3923 and didn't back port it because I thought it was working enough in this PR. Sorry about that. But also, I wasn't sure about fully redefining `sources` like that because I'm new to the project. That makes a lot of sense though and I'm glad for the help.

**Pull request merged**